library/Fission/Storage/IPFS/Add.hs

```diff
@@ -0,0 +1,178 @@
+  IPFS.Proc.run opts raw >>= \case
+    (ExitSuccess, result, _) ->
+      IPFS.Pin.add (mkCID <| UTF8.textShow result) >>= \case
+        Left err ->
+          return . Left . UnknownAddErr <| UTF8.textShow err
+
+        Right _ ->
+          case CL.lines result of
+            [inner, outer] ->
+              let
+                sparseTree  = Directory [(Hash rootCID, fileWrapper)]
+                fileWrapper = Directory [(fileName, Content fileCID)]
+                rootCID     = CID <| UTF8.textShow outer
+                fileCID     = CID . UTF8.stripN 1 <| UTF8.textShow inner
+                fileName    = Key name
+              in
+                return <| Right sparseTree
+
+            bad ->
+              return . Left . UnexpectedOutput <| UTF8.textShow bad
+
+
+    (ExitFailure _, _, err) ->
+      return . Left . UnknownAddErr <| UTF8.textShow err
+
+    where
+      opts = [ "add"
+             , "-wq"
+             , "--stdin-name"
+             , unName name
+             ]
+
+addPath ::
+  ( RIOProc           cfg m
+  , Has IPFS.Timeout  cfg
+  , Has IPFS.BinPath  cfg
+  )
+  => FilePath
+  -> m (Either IPFS.Error.Add CID)
+addPath path = IPFS.Proc.run ["add", "-HQ", path] "" >>= pure . \case
+    (ExitSuccess, result, _) ->
+      case CL.lines result of
+        [cid] -> Right . mkCID . UTF8.stripN 1 <| UTF8.textShow cid
+        bad   -> Left . UnexpectedOutput <| UTF8.textShow bad
+
+    (ExitFailure _, _, err) ->
+      Left . UnknownAddErr <| UTF8.textShow err
+
+addDir ::
+  ( MonadRIO cfg m
+  , HasProcessContext cfg
+  , HasLogFunc cfg
+  , Has IPFS.BinPath cfg
+  , Has IPFS.Timeout cfg
+  )
+  => IPFS.Ignored
+  -> FilePath
+  -> m (Either IPFS.Error.Add IPFS.CID)
+addDir ignored path = doesFileExist path >>= \case
+  True -> addPath path
+  False -> walkDir ignored path
+
+walkDir ::
+  ( MonadRIO cfg m
+  , HasProcessContext cfg
+  , HasLogFunc cfg
+  , Has IPFS.BinPath cfg
+  , Has IPFS.Timeout cfg
+  )
+  => IPFS.Ignored
+  -> FilePath
+  -> m (Either IPFS.Error.Add IPFS.CID)
+walkDir ignored path = do
+  files <- listDirectory path
+  let toAdd = removeIgnored ignored files
+  let f = foldResults path ignored
+  foldM f (Right <| Node { dataBlock = "CAE=", links = [] }) toAdd >>= \case
+    Left err -> return <| Left err
+    Right node -> DAG.putNode node
+
+foldResults ::
+  ( MonadRIO cfg m
+  , HasProcessContext cfg
+  , HasLogFunc cfg
+  , Has IPFS.BinPath cfg
+  , Has IPFS.Timeout cfg
+  )
+  => FilePath
+  -> IPFS.Ignored
+  -> Either IPFS.Error.Add Node
+  -> FilePath
+  -> m (Either IPFS.Error.Add Node)
+foldResults _ _ (Left err) _ = return <| Left err
+foldResults path ignored (Right node) filename = do
+  addDir ignored (path </> filename) >>= \case
+    Left err ->  return <| Left err
+    Right cid ->
+      DAG.Link.create cid (IPFS.Name filename) >>= \case
+      Left err -> return <| Left err
+      Right link ->
+        return <| Right <| node { links = link:(links node) }
```

Comment:
Moving the conversation here from the `Outdated` comment (which is thus hidden). This looks great! Yes, the correct thing to do to "exit early" is

```haskell
foldResults _ _ (Left err) _ = return <| Left err
```